File: backend/src/routes/baidu.js
const express = require('express')
const router = express.Router()
const baiduService = require('../services/baiduService')
const logger = require('../utils/logger')

// 获取访问令牌
router.post('/token', async (req, res) => {
  try {
    const { code, clientId, clientSecret, redirectUri } = req.body
    const token = await baiduService.getTokenViaCode(code, clientId, clientSecret, redirectUri)
    res.json(token)
  } catch (error) {
    logger.error('[API] 获取令牌失败:', error.message)
    res.status(500).json({ error: error.message })
  }
})

// 直接代理百度API获取token（避免CORS）
router.post('/token-direct', async (req, res) => {
  try {
    const { code, clientId, clientSecret, redirectUri } = req.body
    
    // 直接调用百度API
    const axios = require('axios')
    const params = new URLSearchParams()
    params.append('grant_type', 'authorization_code')
    params.append('code', code)
    params.append('client_id', clientId)
    params.append('client_secret', clientSecret)
    params.append('redirect_uri', redirectUri)
    
    const response = await axios.post(
      'https://openapi.baidu.com/oauth/2.0/token',
      params,
      {
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded'
        }
      }
    )
    
    res.json(response.data)
  } catch (error) {
    logger.error('[API] 直接获取令牌失败:', error.message)
    if (error.response) {
      res.status(error.response.status).json(error.response.data)
    } else {
      res.status(500).json({ error: error.message })
    }
  }
})

// 通过alist API获取token
router.post('/alist-token', async (req, res) => {
  try {
    const { code } = req.body
    
    if (!code) {
      return res.status(400).json({ error: '缺少授权码' })
    }
    
    logger.info(`[AlistToken] 通过alist API获取token, code: ${code}`)
    
    // 调用alist API
    const axios = require('axios')
    const response = await axios.get(
      `https://api.alistgo.com/alist/baidu/get_refresh_token?code=${code}`,
      {
        timeout: 30000,
        validateStatus: function (status) {
          return status < 500 // 只有5xx才抛出异常
        }
      }
    )
    
    logger.info('[AlistToken] alist API响应状态:', response.status)
    logger.info('[AlistToken] alist API响应数据:', response.data)
    
    // 检查响应状态
    if (response.status !== 200) {
      logger.error('[AlistToken] alist API返回非200状态:', response.status)
      return res.status(response.status).json({ 
        error: `alist API返回错误状态: ${response.status}`,
        details: response.data 
      })
    }
    
    // 检查响应数据
    if (response.data && response.data.access_token && response.data.refresh_token) {
      logger.info('[AlistToken] 成功获取token')
      res.json(response.data)
    } else {
      logger.error('[AlistToken] alist API返回数据格式错误:', response.data)
      res.status(400).json({ 
        error: 'alist API返回数据格式错误',
        details: response.data 
      })
    }
  } catch (error) {
    logger.error('[AlistToken] 通过alist获取令牌失败:', error.message)
    
    if (error.response) {
      logger.error('[AlistToken] alist API错误响应:', error.response.data)
      return res.status(error.response.status || 500).json({ 
        error: 'alist API请求失败',
        message: error.message,
        details: error.response.data 
      })
    }
    
    if (error.code === 'ECONNABORTED') {
      return res.status(504).json({ error: '请求超时，请稍后重试' })
    }
    
    res.status(500).json({ 
      error: '服务器内部错误',
      message: error.message 
    })
  }
})

// 刷新令牌
router.post('/refresh', async (req, res) => {
  try {
    const { refreshToken, clientId, clientSecret } = req.body
    const newToken = await baiduService.refreshToken(refreshToken, clientId, clientSecret)
    res.json(newToken)
  } catch (error) {
    logger.error('[API] 刷新令牌失败:', error.message)
    res.status(500).json({ error: error.message })
  }
})

// 验证令牌
router.get('/verify', async (req, res) => {
  try {
    const { accessToken } = req.query
    const result = await baiduService.verifyToken(accessToken)
    res.json(result)
  } catch (error) {
    logger.error('[API] 验证令牌失败:', error.message)
    res.status(500).json({ error: error.message })
  }
})

// 获取文件列表
router.get('/files', async (req, res) => {
  try {
    const { accessToken, dir, pageNum, pageSize, order, method, recursion } = req.query
    const files = await baiduService.getFileList(
      accessToken,
      dir,
      parseInt(pageNum) || 1,
      parseInt(pageSize) || 100,
      order || 'name',
      method || 'list',
      parseInt(recursion) || 0
    )
    res.json(files)
  } catch (error) {
    logger.error('[API] 获取文件列表失败:', error.message)
    res.status(500).json({ error: error.message })
  }
})

// 搜索文件
router.get('/search', async (req, res) => {
  try {
    const { accessToken, key, dir, method, recursion } = req.query
    const files = await baiduService.searchFiles(
      accessToken,
      key,
      dir,
      method || 'search',
      parseInt(recursion) || 1
    )
    res.json(files)
  } catch (error) {
    logger.error('[API] 搜索文件失败:', error.message)
    res.status(500).json({ error: error.message })
  }
})

// 获取文件信息
router.get('/fileinfo', async (req, res) => {
  try {
    const { accessToken, fsids } = req.query
    const info = await baiduService.getFileInfo(accessToken, fsids)
    res.json(info)
  } catch (error) {
    logger.error('[API] 获取文件信息失败:', error.message)
    res.status(500).json({ error: error.message })
  }
})

// 下载文件
router.get('/download', async (req, res) => {
  try {
    const { dlink, accessToken } = req.query
    const result = await baiduService.downloadFile(dlink, accessToken)
    res.json(result)
  } catch (error) {
    logger.error('[API] 下载文件失败:', error.message)
    res.status(500).json({ error: error.message })
  }
})

// 上传文件
router.post('/upload', async (req, res) => {
  try {
    const { fileName, fileData, accessToken } = req.body
    
    // fileData 是数组格式，转换为 Buffer
    const fileBuffer = Buffer.from(fileData)
    
    const result = await baiduService.uploadFile(fileName, fileBuffer, accessToken)
    res.json(result)
  } catch (error) {
    logger.error('[API] 上传文件失败:', error.message)
    res.status(500).json({ error: error.message })
  }
})

// 删除文件
router.post('/delete', async (req, res) => {
  try {
    const { accessToken, filePaths } = req.body
    const result = await baiduService.deleteFile(accessToken, filePaths)
    res.json(result)
  } catch (error) {
    logger.error('[API] 删除文件失败:', error.message)
    res.status(500).json({ error: error.message })
  }
})

module.exports = router

File: python-backend/app/routes/baidu.py
"""
Baidu Netdisk API 路由
"""
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional, List
from loguru import logger

from app.services.baidu_service import BaiduNetdiskService

router = APIRouter()
baidu_service = BaiduNetdiskService()


class TokenRequest(BaseModel):
    code: str
    clientId: str
    clientSecret: str
    redirectUri: str


class AlistTokenRequest(BaseModel):
    code: str


class RefreshTokenRequest(BaseModel):
    refreshToken: str
    clientId: str
    clientSecret: str


class FileListRequest(BaseModel):
    accessToken: str
    dir: str = "/"
    pageNum: int = 1
    pageSize: int = 100
    order: str = "name"
    method: str = "list"
    recursion: int = 0


class CreateDirRequest(BaseModel):
    accessToken: str
    dir: str


class UploadFileRequest(BaseModel):
    fileName: str
    fileData: List[int]  # 数组格式的文件数据
    accessToken: str


class DeleteFileRequest(BaseModel):
    accessToken: str
    filePaths: List[str]


@router.post("/token")
async def get_token(request: TokenRequest):
    """获取访问令牌；验证通过后写入统一 token 存储（本地 JSON + 内存缓存）"""
    try:
        token = baidu_service.get_token_via_code(
            request.code,
            request.clientId,
            request.clientSecret,
            request.redirectUri
        )
        if token.get("access_token"):
            from app.services.token_store import set_baidu
            set_baidu(token["access_token"])
        return token
    except Exception as e:
        logger.error(f"获取令牌失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/token-direct")
async def get_token_direct(request: TokenRequest):
    """直接代理百度 API 获取 token（避免 CORS）；成功后写入统一 token 存储"""
    try:
        token = baidu_service.get_token_via_code(
            request.code,
            request.clientId,
            request.clientSecret,
            request.redirectUri
        )
        if token.get("access_token"):
            from app.services.token_store import set_baidu
            set_baidu(token["access_token"])
        return token
    except Exception as e:
        logger.error(f"直接获取令牌失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/alist-token")
async def get_alist_token(request: AlistTokenRequest):
    """通过 alist API 获取 token"""
    try:
        if not request.code:
            raise HTTPException(status_code=400, detail="缺少授权码")
        
        token = baidu_service.get_token_via_alist(request.code)
        if token.get("access_token"):
            from app.services.token_store import set_baidu
            set_baidu(token["access_token"])
        return token
    except Exception as e:
        logger.error(f"通过 alist 获取令牌失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/refresh")
async def refresh_token(request: RefreshTokenRequest):
    """刷新令牌"""
    try:
        new_token = baidu_service.refresh_token(
            request.refreshToken,
            request.clientId,
            request.clientSecret
        )
        if new_token.get("access_token"):
            from app.services.token_store import set_baidu
            set_baidu(new_token["access_token"])
        return new_token
    except Exception as e:
        logger.error(f"刷新令牌失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/verify")
async def verify_token(accessToken: str):
    """验证令牌；验证通过后写入统一 token 存储（本地 JSON + 内存缓存）"""
    try:
        result = baidu_service.verify_token(accessToken)
        from app.services.token_store import set_baidu
        set_baidu(accessToken)
        # 返回前端可识别的成功格式
        return {"success": True, **result} if isinstance(result, dict) else {"success": True, "data": result}
    except Exception as e:
        logger.error(f"验证令牌失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/files")
async def get_files(
    accessToken: str,
    dir: str = "/",
    pageNum: int = 1,
    pageSize: int = 100,
    order: str = "name",
    method: str = "list",
    recursion: int = 0
):
    """获取文件列表"""
    try:
        files = baidu_service.get_file_list(
            accessToken,
            dir,
            pageNum,
            pageSize,
            order,
            method,
            recursion
        )
        return files
    except Exception as e:
        logger.error(f"获取文件列表失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/search")
async def search_files(
    accessToken: str,
    key: str,
    dir: str = "/",
    method: str = "search",
    recursion: int = 1
):
    """搜索文件"""
    try:
        files = baidu_service.search_files(
            accessToken,
            key,
            dir,
            method,
            recursion
        )
        return files
    except Exception as e:
        logger.error(f"搜索文件失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/fileinfo")
async def get_file_info(accessToken: str, fsids: str):
    """获取文件信息"""
    try:
        info = baidu_service.get_file_info(accessToken, fsids)
        return info
    except Exception as e:
        logger.error(f"获取文件信息失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/download")
async def download_file(dlink: str, accessToken: str):
    """下载文件"""
    try:
        result = baidu_service.download_file(dlink, accessToken)
        return result
    except Exception as e:
        logger.error(f"下载文件失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/mkdir")
async def create_directory(request: CreateDirRequest):
    """创建目录"""
    try:
        result = baidu_service.create_directory(
            request.accessToken,
            request.dir
        )
        return result
    except Exception as e:
        logger.error(f"创建目录失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/upload")
async def upload_file(request: UploadFileRequest):
    """上传文件"""
    try:
        # 将数组转换为 bytes
        file_buffer = bytes(request.fileData)
        
        result = baidu_service.upload_file(
            request.fileName,
            file_buffer,
            request.accessToken
        )
        return result
    except Exception as e:
        logger.error(f"上传文件失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/delete")
async def delete_file(request: DeleteFileRequest):
    """删除文件"""
    try:
        result = baidu_service.delete_file(
            request.accessToken,
            request.filePaths
        )
        return result
    except Exception as e:
        logger.error(f"删除文件失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))
